src/test/java/org/sagebionetworks/bridge/services/AssessmentServiceTest.java

```diff
@@ -0,0 +1,1051 @@
+        Assessment assessment = AssessmentTest.createAssessment();
+        service.createAssessment(APP_ID_VALUE, assessment);
+    }
+    
+    @Test(expectedExceptions = EntityAlreadyExistsException.class)
+    public void createAssessmentAlreadyExists() {
+        Assessment assessment = AssessmentTest.createAssessment();
+        assessment.setGuid(null);
+        assessment.setDeleted(false);
+        
+        when(mockSubstudyService.getSubstudy(APP_AS_STUDY_ID, OWNER_ID, false))
+            .thenReturn(mockSubstudy);
+        when(mockDao.getAssessmentRevisions(any(), any(), anyInt(), anyInt(), anyBoolean()))
+            .thenReturn(new PagedResourceList<>(ImmutableList.of(assessment), 1));
+    
+        service.createAssessment(APP_ID_VALUE, assessment);
+    }
+    
+    @Test(expectedExceptions = InvalidEntityException.class,
+            expectedExceptionsMessageRegExp = ".*identifier cannot be missing.*")
+    public void createAssessmentInvalid() {
+        when(mockSubstudyService.getSubstudy(APP_AS_STUDY_ID, OWNER_ID, false))
+            .thenReturn(mockSubstudy);
+        when(mockDao.getAssessmentRevisions(any(), any(), anyInt(), anyInt(), anyBoolean()))
+            .thenReturn(EMPTY_LIST);
+        
+        Assessment assessment = AssessmentTest.createAssessment();
+        assessment.setIdentifier(null);
+        
+        service.createAssessment(APP_ID_VALUE, assessment);
+    }
+    
+    @Test
+    public void createAssessmentScrubsMarkup() {
+        when(mockSubstudyService.getSubstudy(APP_AS_STUDY_ID, OWNER_ID, false))
+            .thenReturn(mockSubstudy);
+        when(mockDao.getAssessmentRevisions(any(), any(), anyInt(), anyInt(), anyBoolean()))
+            .thenReturn(EMPTY_LIST);
+        
+        Assessment assessment = AssessmentTest.createAssessment();
+        addMarkupToSensitiveFields(assessment);
+
+        service.createAssessment(APP_ID_VALUE, assessment);
+        
+        assertMarkupRemoved(assessment);
+    }
+
+    @Test
+    public void createAssessmentRevision() {
+        when(mockSubstudyService.getSubstudy(APP_AS_STUDY_ID, OWNER_ID, false))
+                .thenReturn(mockSubstudy);
+        when(mockDao.getAssessment(APP_ID_VALUE, GUID))
+            .thenReturn(Optional.of(AssessmentTest.createAssessment()));
+        
+        Assessment assessment = AssessmentTest.createAssessment();
+        assessment.setGuid(null);
+        assessment.setDeleted(true); // can't do this, it's reset
+        
+        service.createAssessmentRevision(APP_ID_VALUE, GUID, assessment);
+        
+        verify(mockDao).saveAssessment(APP_ID_VALUE, assessment);
+        
+        assertEquals(assessment.getGuid(), GUID);
+        assertEquals(assessment.getOwnerId(), OWNER_ID);
+        // same timestamp on creation
+        assertEquals(assessment.getCreatedOn(), CREATED_ON);
+        assertEquals(assessment.getModifiedOn(), CREATED_ON);
+        assertFalse(assessment.isDeleted());
```

Comment:
Verify that we can't change the identifier?

Reply:
Added